lektor/packages.py

```diff
@@ -74,251 +87,205 @@ def remove_package_from_project(project, name):
-
-
-def list_local_packages(path):
-    """Lists all local packages below a path that could be installed."""
-    rv = []
-    try:
-        for filename in os.listdir(path):
-            if os.path.isfile(os.path.join(path, filename, "setup.py")):
-                rv.append("@" + filename)
-    except OSError:
-        pass
-    return rv
-
-
-def update_cache(package_root, remote_packages, local_package_path, refresh=False):
-    """Updates the package cache at package_root for the given dictionary
-    of packages as well as packages in the given local package path.
+class VirtualEnv:
+    """A helper for manipulating our private package cache virtual environment.
+
+    Parameters:
+
+    path — The path to the virtual environment to manage.  This can be an existing
+    environment or not.
+
     """
-    requires_wipe = False
-    if refresh:
-        click.echo("Force package cache refresh.")
-        requires_wipe = True
 
-    manifest_file = os.path.join(package_root, "lektor-packages.manifest")
-    local_packages = list_local_packages(local_package_path)
+    def __init__(self, path: StrPath):
+        self.path = Path(path)
+
+    def create(self, with_pip: bool = True, upgrade_deps: bool = True) -> None:
+        """(Re-)Create a new virtual environment.
+
+        This will remove any existing virtual environment and create a new one.
+
+        The parameters ``with_pip`` and ``upgrade_deps`` should probably be left at
+        their default values in normal usage.  They are provided here mostly for use in
+        tests. They work as described for ``venv.EnvBuilder`` from the standard library.
+        (Though ``upgrade_deps`` is only supported by EnvBuilder`` in py39+, here we
+        emulate it's behavior if running under older pythons.)
+
+        """
+        # Right now, by default, we always install and upgrade pip to
+        # the latest available version.
+        #
+        # We could optimize by not installing (and not upgrading) pip if
+        # the system pip is sufficient to our needs.
+        #
+        # Note that, e.g., pip>=21.3 is required to support PEP660 editable
+        # installs.
+        options: Dict[str, Any] = {"clear": True, "with_pip": with_pip}
+        if sys.version_info >= (3, 9):
+            EnvBuilder(upgrade_deps=upgrade_deps, **options).create(self.path)
+        else:
+            EnvBuilder(**options).create(self.path)
+            if upgrade_deps:
+                self.run_pip_install("--upgrade", "pip", "setuptools")
+
+    def addsitedir(self, sitedir: str) -> None:
+        """Add an additional sitedir to sys.path for virtual environment.
+
+        Packages installed in ``sitedir`` will be made available to any invocations
+        of python running within the virtual environment.
+        """
+        with Path(self.site_packages, "_lektor.pth").open("a", encoding="utf-8") as fp:
+            fp.write(f"import site; site.addsitedir({sitedir!r})\n")
```

Comment:
```suggestion
            fp.write(sitedir)
```
Simply writing the path to the `.pth` file would work.

>A path configuration file is a file whose name has the form `name.pth` and exists in one of the four directories mentioned above; its contents are additional items (one per line) to be added to `sys.path`.

Reply:
In addition to adding the directory to `sys.path`, `site.addsitedir` also processes any `.pth` files found therein.
We need that here, and I'm pretty sure that doesn't happen if we just list the directory in the `.pth` file.

Reply:
Indeed I couldn't get the point adding lektor site-packages to the venv. However, the venv packages must be loaded back into lektor's sys.path.